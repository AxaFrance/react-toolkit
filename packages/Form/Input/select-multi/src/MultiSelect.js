import React from 'react';
import PropTypes from 'prop-types';
import ReactSelect from 'react-select';

import {
  Input,
  InputConstants as Constants,
  withInput,
  omit,
} from '@axa-fr/react-toolkit-form-core';

const omitProperties = omit(['classModifier', 'className', 'isVisible']);

const MultiSelect = props => {
  const {
    name,
    loadOptions,
    values,
    options,
    value,
    onChange,
    onBlur,
    placeholder,
    className,
    readOnly,
    disabled,
    loadingPlaceholder,
    searchPromptText,
    noResultsText,
    ...otherProps
  } = props;
  
  const isDisabled = disabled || readOnly;
  const SelectComponent = loadOptions ? ReactSelect.Async : ReactSelect;

  if (values != null) {
    const newValues = options.reduce((acc, opt) => {
      const valueTemp = values.find(v => v === opt.value);
      if (valueTemp) {
        acc.push(opt);
      }
      return acc;
    }, []);

    return (
      <SelectComponent
        isMulti
        name={name}
        value={newValues}
        onChange={onChange}
        onBlur={onBlur}
        placeholder={placeholder}
        className={className}
        isDisabled={isDisabled}
        options={options}
        loadOptions={loadOptions}
        loadingPlaceholder={loadingPlaceholder}
        searchPromptText={searchPromptText}
        noResultsText={noResultsText}
        valueKey="value"
        labelKey="label"
        {...omitProperties(otherProps)}
      />
    );
  }
  const newValue = options.find(v => v.value === value) || '';
  return (
    <SelectComponent
      multi={false}
      name={name}
      value={newValue}
      placeholder={placeholder}
      onChange={onChange}
      onBlur={onBlur}
      className={className}
      isDisabled={isDisabled}
      options={options}
      loadOptions={loadOptions}
      loadingPlaceholder={loadingPlaceholder}
      searchPromptText={searchPromptText}
      noResultsText={noResultsText}
      valueKey="value"
      labelKey="label"
      {...omitProperties(otherProps)}
    />
  );
};

const propTypes = {
  ...Constants.propTypes,
  values: PropTypes.array,
  options: PropTypes.array.isRequired,
  placeholder: PropTypes.string,
  loadingPlaceholder: PropTypes.string,
  searchPromptText: PropTypes.string,
  noResultsText: PropTypes.string,
  loadOptions: PropTypes.func,
};

const defaultClassName = 'react-select';
const defaultProps = {
  ...Constants.defaultProps,
  values: null,
  placeholder: 'Select',
  loadingPlaceholder: 'Chargement...',
  searchPromptText: 'Saisir pour chercher',
  noResultsText: 'Aucun rÃ©sultat',
  loadOptions: null,
  className: defaultClassName,
};

const handlers = {
  onChange: ({ values, name, id, onChange }) => newValue => {
    if (values != null) {
      const newValues = newValue.map(v => v.value);
      onChange({
        values: newValues,
        name,
        id,
      });
    } else {
      onChange({
        value: newValue.value,
        name,
        id,
      });
    }
  },
};

const EnhancedComponent = withInput(
  defaultClassName,
  propTypes,
  defaultProps,
  handlers
)(MultiSelect);

EnhancedComponent.Clone = Input.Clone;
EnhancedComponent.displayName = MultiSelect.name;

export default EnhancedComponent;
