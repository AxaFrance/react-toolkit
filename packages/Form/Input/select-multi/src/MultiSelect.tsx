import React from 'react';
import ReactSelectAsync from 'react-select/async';
import { CustomFormEventValues } from '@axa-fr/react-toolkit-form-core';
import {
  OptionsType,
  OptionTypeBase,
  NamedProps as LibProps,
  ValueType,
} from 'react-select';

type Props = Omit<
  LibProps<OptionTypeBase>,
  'onChange' | 'loadOptions' | 'value'
> & {
  readOnly?: boolean;
  disabled?: boolean;
  onChange?: (event: CustomFormEventValues) => void;
  values?: string[];
  loadOptions?: (
    inputValue: string,
    callback: (options: OptionsType<OptionTypeBase>) => void
  ) => Promise<any> | void;
};

const MultiSelect = ({
  id,
  name,
  loadOptions,
  values,
  options,
  onChange,
  readOnly,
  disabled,
  placeholder = 'Select',
  className = 'react-select',
  noOptionsMessage = () => 'Aucun rÃ©sultat',
  ...otherProps
}: Props) => {
  const selectedValues = options.filter((opt) => values?.includes(opt.value));

  const loadOptionsDefault = () => Promise.resolve(options);

  const handleChange = (changedValue: ValueType<OptionTypeBase>) => {
    onChange &&
      onChange({
        id,
        name,
        values: (changedValue || []).map((v: OptionTypeBase) => v.value),
      });
  };
  return (
    <ReactSelectAsync
      {...otherProps}
      className={className}
      isMulti={values !== null}
      name={name}
      value={selectedValues}
      placeholder={placeholder}
      noOptionsMessage={noOptionsMessage}
      onChange={handleChange}
      isDisabled={disabled || readOnly}
      defaultOptions
      loadOptions={loadOptions || loadOptionsDefault}
      valueKey="value"
      labelKey="label"
    />
  );
};

export default MultiSelect;
