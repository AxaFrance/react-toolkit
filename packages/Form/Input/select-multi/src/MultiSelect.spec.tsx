import React from 'react';
import renderer from 'react-test-renderer';
import MultiSelect from './MultiSelect';

export const options = [
  { value: 'fun', label: 'For fun', clearableValue: false },
  { value: 'work', label: 'For work' },
  { value: 'drink', label: 'For drink' },
  { value: 'sleep', label: 'For sleep' },
  { value: 'swim', label: 'For swim' },
];

export const values = ['fun', 'drink'];

describe('<MultiSelect>', () => {
  it('renders MultiSelect correctly', () => {
    const tree = renderer
      .create(
        <MultiSelect
          name="placeName"
          onChange={() => null}
          options={options}
          values={values}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
