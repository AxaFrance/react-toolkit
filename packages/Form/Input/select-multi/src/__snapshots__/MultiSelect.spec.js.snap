// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSelect> renders MultiSelect correctly 1`] = `
<div
  className="css-10nd86i react-select"
  onKeyDown={[Function]}
>
  <div
    className="css-vj8t7z"
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="css-1hwfws3"
    >
      <div
        className="css-xwjg1b"
      >
        <div
          className="css-12jo7m5"
        >
          For fun
        </div>
        <div
          className="css-1alnv5e"
          onClick={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-19bqh2r"
            focusable="false"
            height={14}
            viewBox="0 0 20 20"
            width={14}
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
      </div>
      <div
        className="css-xwjg1b"
      >
        <div
          className="css-12jo7m5"
        >
          For drink
        </div>
        <div
          className="css-1alnv5e"
          onClick={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-19bqh2r"
            focusable="false"
            height={14}
            viewBox="0 0 20 20"
            width={14}
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
      </div>
      <div
        className="css-1g6gooi"
      >
        <div
          className=""
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-autocomplete="list"
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            disabled={false}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "boxSizing": "content-box",
                "color": "inherit",
                "fontSize": "inherit",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "1px",
              }
            }
            tabIndex="0"
            type="text"
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-1wy0on6"
    >
      <div
        aria-hidden="true"
        className="css-1ep9fjw"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <svg
          aria-hidden="true"
          className="css-19bqh2r"
          focusable="false"
          height={20}
          viewBox="0 0 20 20"
          width={20}
        >
          <path
            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
          />
        </svg>
      </div>
      <span
        className="css-d8oujb"
      />
      <div
        aria-hidden="true"
        className="css-1ep9fjw"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <svg
          aria-hidden="true"
          className="css-19bqh2r"
          focusable="false"
          height={20}
          viewBox="0 0 20 20"
          width={20}
        >
          <path
            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div>
    <input
      name="placeName"
      type="hidden"
      value="fun"
    />
    <input
      name="placeName"
      type="hidden"
      value="drink"
    />
  </div>
</div>
`;
